packages/docusaurus/docs/features/constraints.mdx

```diff
@@ -125,37 +125,251 @@ This section regroups a couple of constraint examples. We are thinking to provid
+  return range;
+}
+
+// Require that if two workspaces have the same dependency, the greater version must be used in both.
+function enforceConsistentDependencies(Yarn) {
+  // Force consistent dependency versions of all dependencies and devDependencies
+  for (const dependency of Yarn.dependencies()) {
+    for (const otherDependency of Yarn.dependencies({
+      ident: dependency.ident,
+    })) {
+      if (dependency.range !== otherDependency.range) {
+        dependency.update(
+          preferredRange(dependency.range, otherDependency.range),
+        );
+      }
+    }
+  }
+}
+
+module.exports = defineConfig({
+  async constraints({ Yarn }) {
+    enforceConsistentDependencies(Yarn);
+  },
+});
+
+```
+
+### Prevent multiple virtual instances of packages
+
+Having multiple instances can break many popular packages like react-dom, styled-components, and so on.
+This constraint enforces that peer dependencies of packages are satisfied in the same way, so that there
+will only be a single instance.  To allow packages to have multiple instances, add them to the
+multipleInstancesWhitelist array.
+
+```ts
+// @ts-check
+
+/** @type {import('@yarnpkg/types')} */
+const { defineConfig } = require('@yarnpkg/types');
+const semver = require('semver');
+
+// If otherRange is a greater version, return it.  Otherwise return range.
+function preferredRange(range, otherRange) {
+  if (otherRange === '*') {
+    return range;
+  }
+  const version = semver.coerce(range);
+  const otherVersion = semver.coerce(otherRange);
+  if (
+    version && otherVersion && !semver.eq(version, otherVersion)
+      ? semver.gt(otherVersion, version)
+      : range.localeCompare(otherRange) > 0
+  ) {
+    return otherRange;
+  }
+  return range;
+}
+
+// Having multiple instances can break many popular packages like react-dom, styled-components, and so on.
+// This constraint enforces that peer dependencies of packages are satisfied in the same way, so that there
+// will only be a single instance.  To allow packages to have multiple instances, add them to the
+// multipleInstancesWhitelist list here.
+const multipleInstancesWhitelist = [
+  'debug',
+  'isomorphic-ws',
+  /^@babel/,
+  /^@types/,
+];
+function enforceSingletonPackagesWithPeerDependencies(Yarn) {
+  // Whitelist deps we don't mind having multiple instances
+  // Force all dependencies to be singletons
+  const inverseDepTree = new Map();
+  const walked = new Set();
   for (const dependency of Yarn.dependencies()) {
-    if (dependency.type === `peerDependencies`)
+    // We don't consider duplicate dev dependencies to be an issue; only multiple instances
+    // of regular or peer dependencies cause problems
+    if (dependency.type === 'devDependencies') {
       continue;
+    }
+    // Can't do much if the dependency failed to resolve
+    if (!dependency.resolution) {
+      continue;
+    }
+    const inverseDepNode = package => {
+      const existing = inverseDepTree.get(package.ident);
```

Comment:
False negative here and in `walk` below, because
- the same package ident can appear multiple times in the dependency tree
- just because two packages have the same ident and version does not mean they are the same package (e.g. one comes from `npm` and one is a local workspace). AFAIK, the data needed to determine this is not available to constraints.